docs/spec/fastsync/fastsync.md

```diff
@@ -0,0 +1,1101 @@
+----
+
+
+```go
+func OnBlockResponse(addr Address, b Block)
+```
+- Comment
+    - if after adding block *b*, blocks of heights *height* and
+      *height + 1* are in *blockstore*, then `Execute` is called
+- Expected precondition
+    - *pendingblocks(b.Height) = addr*
+	- *b* satisfies basic soundness  
+- Expected postcondition
+    - if function `Execute` has been executed without error or was not executed:
+        - *receivedBlocks(b.Height) = addr*
+	    - *blockstore(b.Height) = b*
+		- *peerTimeStamp[addr]* is set to a time between invocation and
+          return of the function.
+		- *peerRate[addr]* is updated according to size of received
+          block
+- Error condition
+    - if precondition is violated: *addr* not in *peerIDs*; reset
+	*pendingblocks(b.Height)* to nil;
+- Timeout condition
+    - if `OnBlockResponse(addr, b)` was not invoked within *2 Delta* after
+	`Block(addr,h)` was called for *b.Height = h*: *addr* not in *peerIDs*
+----
+
+
+
+```go
+func Execute()
+```
+- Comments
+    - none
+- Expected precondition
+	- application state is the one of the blockchain at height
+      *height - 1*
+	- **[FS-V2-Verif]** for any two blocks *a* and *b* from
+	*receivedBlocks*: if
+	  *a.Height + 1 = b.Height* then *VerifyCommit (a,b.Commit) = true*
+- Expected postcondition
+    - Any two blocks *a* and *b* violating [FS-V2-Verif]:
+	  *a* and *b* not in *blockstore*; nodes with Address 
+	  receivedBlocks(a.Height) and receivedBlocks(b.Height) not in peerIDs
+	- height is updated height of complete prefix that matches the blockchain
+	- state is the one of the blockchain at height *height - 1*
+- Error condition
+    - none
+----
+
+## Analysis of Fastsync V2
+
+####  **[FS-ISSUE-KILL]**:
+If two blocks are not matching [FS-V2-Verif], `Execute` dismisses both
+blocks and removes the peers that provided these blocks from
+*peerIDs*. If block *a* was correct and provided by a correct peer *p*,
+and block b was faulty and provided by a faulty peer, the protocol
+- removes the correct peer *p*, although it might be useful to
+  download blocks from it in the future
+- removes the block *a*, so that a fresh copy of *a* needs to be downloaded
+  again from another peer
+  
+By [FS-A-PEER] we do not put a restriction on the number
+  of faulty peers, so that faulty peers can make *FS* to remove all
+  correct peers from *peerIDs*. As a result, this version of
+  *Fastsync* violates [FS-VC-CORR-INV-SYNC].
+
+
+####  **[FS-ISSUE-NON-TERM]**:
+
+Due to [**[FS-ISSUE-KILL]**](#fs-issue-kill), from some point on, only
+faulty peers may be in *peerIDs*. They can thus control at which rate
+*Fastsync* gets blocks. If the timeout duration from [FS-V2-TIMEOUT]
+is greater than the time it takes to add a block to the blockchain
+(LTIME in [**[TMBC-SEQ-APPEND-E]**][TMBC-SEQ-APPEND-E-link]), the
+protocol may never terminate and thus violate [FS-VC-TERM].  This
+scenario is even possible if a correct peer is always in *peerIDs*,
```

Comment:
Slash isn't it possible with just one correct peer if we manage to always eventually request pairs of sequential blocks from that correct peer?

Reply:
I guess your comment applies to the case above: we could ask always consecutive blocks from a peer, but it would lead to waste of resources as you would need for every block to download two blocks (in case requests are made to different peers). More efficient solution would be asking block and commit from a single peer, but that would require Tendermint API.